Update audit-log.rst (#5083)
mattermost/docs: source/comply/audit-log.rst

System Admins can define multiple log targets to:

- Mirror log output to files and log aggregators for redundancy.
- Log certain entries to specific destinations. For example, all errors could be routed to a specific destination for review.

All access to the REST API or CLI is audited. When using Advanced Logging for auditing, System Admins can capture the following auditing in the target configuration in addition to discrete log levels:

.. code-block:: none

   "Levels": [
      {"ID": 100, "Name": "audit-api"},
      {"ID": 101, "Name": "audit-content"},
      {"ID": 102, "Name": "audit-permissions"},
      {"ID": 103, "Name": "audit-cli"},
   ],

Where:

- ``audit-api``: Enables output of REST API calls.
- ``audit-content``: Enables output of API calls that generate content (e.g. ``create post``, ``create reaction``).
- ``audit-permissions``: Enables output of all permissions failures.
- ``audit-cli``: Enables output of legacy CLI calls.

All levels can be viewed at ``mattermost-server/shared/mlog/levels.go``.

.. Note::
  - Logs are recorded asynchronously to reduce latency to the caller. 

>   - Advanced logging supports hot-reloading of logger configuration.

Reviewer: Hot-reloading of all config files was removed from V6 server, including audit.

Author: Will remove this section. Thx for calling out.

Author: Actually will add a note for 6.0 for accuracy.

File target supports rotation and compression triggered by size and/or duration. Syslog target supports local and remote syslog servers, with or without TLS transport. TCP socket target can be configured with an IP address or domain name, port, and optional TLS certificate.

.. note::
   Mattermost v6.0 introduced a change to the logging engine library which changes how the logger is configured in Mattermost 5.x versions. Please refer to :download:`Advanced Logging sample file <../samples/sample-logger-config.json>` for a sample configuration file supported in in v6.0 and beyond. 

   
**Accessing configuration options for audit log**

Open ``config.json`` and navigate to the audit settings under ``ExperimentalAuditSettings``. Within the setting ``AdvancedLoggingConfig`` you are able to specifiy a filespec to another config file, a database DSN, or JSON. 

The logging configuration JSON is an object (unordered collection) containing names and log target values. Each log target contains a type, options specific to the target type, format, and levels.

> The example below specifies one log target that outputs to the console using a plain text format with pipes delimiting fields. The standard log levels are listed, with only `error` and lower outputing a stack trace. The `error` level will output in red for log targets and formatters that support color output.

Reviewer: ```suggestion
The example below specifies one log target that outputs to the console using a plain text format with pipes delimiting fields. The standard log levels are listed, with only `error` and lower outputting a stack trace. The `error` level will output in red for log targets and formatters that support color output.
```

.. code-block:: json

  {
      "sample-console": {
        "type": "console",
        "options": {
          "out": "stdout"
        },
        "format": "plain",
        "format_options": {
          "delim": " | "
        },
        "levels": [
          {"id": 5, "name": "debug"},
          {"id": 4, "name": "info"},
          {"id": 3, "name": "warn"},
          {"id": 2, "name": "error", "stacktrace": true, "color": 31},
          {"id": 1, "name": "fatal", "stacktrace": true},
          {"id": 0, "name": "panic", "stacktrace": true}
        ],
        "maxqueuesize": 1000
      }
    }
    

> .. Note::

Reviewer 2: ```suggestion
.. note::
```

    Filenames for ``AdvancedLoggingConfig`` can contain an absolute filename, a relative filename, or embedded JSON.

See the :download:`Advanced Logging Options Sample JSON ZIP file <../samples/advanced-logging-options-sample-json.zip>` for a sample configuration file.  

Log Target Types
~~~~~~~~~~~~~~~~
Log target types include Console, Syslog, File, or TCP. 

> **Console configuration options:**

Reviewer 2: These sections may be good candidates for using a tabbed content approach, though I'm not 100% sure that CSV-based tables will display as intended. The basic format goes like this:

```
.. tabs::
   
    .. tab:: tabname

    .. tab:: tabname

    .. tab:: tabname
```

I'm happy to help format this content as tabs and see if CSV-based tables display as intended. Let me know if you're open to this approach.

Author: Let's do the tabbed content post 6.0 when we re-vamp the configuration sections.  I am sure this one will get re-reviewed at that time.

.. csv-table::
    :header: "Key", "Type", "Default", "Description"
       
       "out", "string", " ", "One of `stdout` or `stderr`. "

**Syslog configuration options:**

.. csv-table::
    :header: "Key", "Type", "Default", "Description"
       
       "host", "string", " ", "IP or domain name of server to receive log records."
       "port", "number", "", "Port number for server receiving log records."
       "tls", "bool", "false", "When true, a TLS connection will be created."
       "cert", "string", "  ", "Path to a cert file (.pem) to be used when establishing a TLS connection."

>        "insecure", "bool", "false", "When true, Focalbaord will accept any certificate presented by the server and any host name in that certificate. Should be used only in testing environments."

Reviewer: ```suggestion
       "insecure", "bool", "false", "When true, Boards will accept any certificate presented by the server and any host name in that certificate. Should be used only in testing environments."
```

Reviewer: Are these the board docs or channels?

Author: @wiggin77 These are outside of channels - Audit log is under administration.  I see this more of a "platform" type feature, so wanted to remove it from the verticals specifically.

Reviewer: Thanks @wiersgallak. If these docs are not product specific shouldn't we leave `Boards` product name out of these config docs?  The same docs apply to channels and boards.

Reviewer: The accepted suggestion below works here:
```
"insecure", "bool", "false", "When true, any certificate will be accepted and any host name in that certificate. Should be used only in testing environments."
```

**File configuration options:**

.. csv-table::
    :header: "Key", "Type", "Default", "Description"

       "filename", "string", "  ", "Full path to output file."
       "max_size", "number", "100", "Maximum size in megabytes the log file can grow before it gets rotated."
       "max_age", "number", "0", "Maximum number of days to retain old log files based on the timestamp encoded in their filename. 0 means do not remove old log files based on age."
       "max_backups", "number", "0", "Maximum number of old log files to retain.  0 means retain all old log files (though max_age may still cause them to get deleted.)"
       "compress", "bool", "false", "When true, the rotated log files will be compressed using gzip."

    
**TCP configuration options:** 

.. csv-table::
    :header: "Key", "Type", "Default", "Description"
       
       "host", "string", "  ", "IP or domain name of server to receive log records."
       "port", "number", " ", "Port number for server receiving log records."
       "tls", "bool", "false", "When true, a TLS connection will be created."
       "cert", "string", "  ", "Path to a cert file (.pem) to be used when establishing a TLS connection."

>        "insecure", "bool", "false", "When true, Mattermost will accept any certificate presented by the server and any host name in that certificate. Should be used only in testing environments."

Reviewer: `Mattermost` is mentioned here, and `FocalBoard` above. Perhaps this section can be made generic:

```suggestion
       "insecure", "bool", "false", "When true, any certificate will be accepted and any host name in that certificate. Should be used only in testing environments."
```

    
To temporarily disable a log target its type can be set to "none".

Log Formatters
~~~~~~~~~~~~~~~~~~~

> Mattermost currently supports three log formats: plain, json, and `GELF <https://docs.graylog.org/en/4.0/pages/gelf.html>`_.

Reviewer 2: ```suggestion
Mattermost currently supports three log formats: plain, JSON, and `GELF <https://docs.graylog.org/en/4.0/pages/gelf.html>`__.
```

**Plain configuration options:**

.. csv-table::
    :header: "Key", "Type", "Default", "Description"
 
       "disable_timestamp", "bool", "alse", "Disables output of the timestamp."
       "disable_level", "bool", "false", "Disables output of the level name."
       "disable_msg", "bool", "false", "Disables output of the message text."
       "disable_fields", "bool", "false", "Disables output of all fields."
       "disable_stacktrace", "bool", "false", "Disables output of stack traces."
       "delim", "string", "single space", "Delimiter placed between fields."
       "min_level_len", "number", "0", "Minimum level name length. If the level name is less than the minimum it will be padded with spaces."
       "min_msg_len", "number", "0", "Minimum msg length. If the msg text is less than the minimum it will be padded with spaces."
       "timestamp_format", "string", "2006-01-02 15:04:05.000 Z07:00", "Format for timestamps. See `format <https://golang.org/pkg/time/#Time.Format>`_ for format details."
       "line_end", "string", "\n ", "Alternative end of line character(s)."
       "enable_color", "bool", "false", "Enables color for targets that support color output."

> **Json configuration options:**

Reviewer 2: ```suggestion
**JSON configuration options:**
```

When using remote syslog, the current best practice is to also write to local file so no records are lost. Note that this does not automatically take records from local file and send it to syslog when syslog becomes available again.

> Configure audit log in Focalboard

Reviewer 2: ```suggestion
Configure audit log in Boards
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> The `Focalboard configuration file <https://github.com/mattermost/focalboard/blob/main/config.json>`_ ``config.json`` is used to configure logging.

Reviewer 2: ```suggestion
The `Boards configuration file <https://github.com/mattermost/focalboard/blob/main/config.json>`_ ``config.json`` is used to configure logging.
```

``logging_cfg_file`` is used to specify the path to a file containing the logging configuration in JSON format.

``logging_cfg_json`` is used to provide logging configuration directly as embedded JSON. Typically this is overridden using the corresponding environment variable ``FOCALBOARD_LOGGING_CFG_JSON``.

> Both configuration methods can be used, but care must be taken to avoid mutiple log targets writing to the same file.

Reviewer: ```suggestion
Both configuration methods can be used, but care must be taken to avoid multiple log targets writing to the same file.
```

The logging configuration JSON is an object (unordered collection) containing names and log target values. Each log target contains a type, options specific to the type, format, and levels.

> Focalboard uses discrete log levels, meaning each level to be output must be listed. This allows for log targets to output specific log levels, and custom log levels to be created. See ``server/mlog/levels.go`` for a list of available log levels. 

Reviewer 2: ```suggestion
Boards uses discrete log levels, meaning each level to be output must be listed. This allows for log targets to output specific log levels, and custom log levels to be created. See ``server/mlog/levels.go`` for a list of available log levels. 
```

> - Allow discrete logging levels. Currently, an application-wide logging level is configured and any log records matching that level or lower will be emitted. These logging levels will remain, but support for zero or more discrete logging levels will be added, meaning only records matching the current log level or one of the discrete levels are emitted. Within the logging engine, any level below 10 (``trace`` through ``critical``/``fatal``, plus ``reserved``) will behave as it does currently, but any level above 10 will be considered discrete. Audit records will have a level above 10.

Reviewer: This does not apply to auditing.

Author: Will remove this line.  Thanks.

> - Allow logging levels and discrete levels to different targets (files, databases, etc) via configuration.

Reviewer: This also does not apply to auditing, and is in place for general logging and no longer in the planning stage.

Author: @wiggin77 Just to confirm you are advocating to remove "- Allow logging levels and discrete levels to different targets (files, databases, etc) via configuration." (line 81) in addition to line 79?

Reviewer: All three bullet points above have been implemented and are in product.  Line 77, 79, 81.
